kubernetes/controllers/pod.go

```diff
@@ -113,31 +113,14 @@ func SetClusterSpecForPod(job interface{}, podTemplate *corev1.PodTemplateSpec)
 		maxReplicas = desiredReplicas
 	}
 
+	launchDefaultArgs := []string{
+		"--rdzv_backend=etcd",
+		"--rdzv_endpoint=" + elasticJob.Spec.RdzvEndpoint,
+		"--rdzv_id=" + elasticJob.Name,
+		"--nnodes=" + strconv.Itoa(int(minReplicas)) + ":" + strconv.Itoa(int(maxReplicas))}
```

Comment:
@kiukchung 

Trying to understand the user experience here. The original plan is user need to load env to get setting and construct the object. Not sure if `torchelastic/distributed/launch.py` will alway be used as user's model entry point. We at least need to let user parse these args in their applications.

Reply:
`torchelastic.distributed.launch` is very similar to `torch.distributed.launch`. It does some setup work and for us it creates an elastic-agent process (which in turn creates `n` subprocesses to start the user script). So the ENTRYPOINT to the docker is `torchelastic.distributed.launch` and the args  shown above are set by the controller, but then the rest of the args are obtained from the spec yaml 

See details on usage here: http://pytorch.org/elastic/0.2.0rc0/distributed.html

Reply:
@Jeffwan  BTW do you know how I can get the job's "run_id" rather than the name in 
```
"--rdzv_id=" + elasticJob.Name
```
the problem with setting the `rdzv_id` as the name of the job is that between two consecutive runs of the same job, it will get the same `rdzv_id` which means that our etcd_rdzv will go to the same directory in etcd to look for the state. So in effect this fools the rdzv algorithm to think that the second run is a "continuation" of the first run.

Reply:
Got you. If you want to make it kind of stateless. I think we can use `uid` here. Each kubernetes object will have a system-generated uid.

https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids

API. 
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#objectmeta-v1-meta

An example string is `uid: 2441e881-76ff-11ea-8085-021721b51cb6`

We can change to `"--rdzv_id=" + elasticJob.Name + elasticJob.ObjectMeta.UID` Not sure limitation on etcd key, may need to truncation

Reply:
Thanks @Jeffwan. I just realized that if we add this we can't support resizes via `kubectl apply -k config/samples/imagenet.yaml` since once I edit the spec yaml k8s will generate a new UID, hence the scaled up nodes will end up creating their own rdzv quorum.

If you have any ideas to make resizes work, but distinguish them from "new runs" let me know.

Another thing I wanted to pick your brain on was to see if we can automate getting the ectd server endpoint if the `rdzvEndpoint` field is left empty. What we can do is the following:

1. add a field in `ElasticJob` called `rdzvService` (defaults to `etcd-service`)
2. programmatically query the namespace `elastic-job` for a service with the name equal to the one specified in `rdzvService`. 
3. grab the cluster-ip (if there are many, round-robin) and port
4. use this `ip:port` as `rdzvEndpoint` if the user did not explictly set one.

I wasn't sure where the best place to add ^^^ was.

Reply:
@kiukchung Doc is not clear on this. 

> It is intended to distinguish between historical occurrences of similar entities.

entities means the different objects with same name, but not different revision of same object. There's a `resourceVersion` field used to indicate the version for same `uid`.

em. `rdzvEndpoint` is actually a `rdzvService`. User needs to pass a k8s service with opening port. trying to understand the use case here. 

> we can automate getting the ectd server endpoint if the rdzvEndpoint field is left empty

Do you want to create an etcd server for that job or auto-discover `elastic-job` namespace to find some potential service to use? 
If that's second one, I think it's unnecessary to predicate the service if user just forget to specify it in the yaml. 

Creating job based etcd if `rdzvEndpoint` missing is a very reasonable feature request and we talked about this before but didn't add it in init version. We can add it.

Reply:
@kiukchung Doc is not clear on this. 

> It is intended to distinguish between historical occurrences of similar entities.

entities means the different objects with same name, but not different revision of same object. There's a `resourceVersion` field used to indicate the version for same `uid`.

em. `rdzvEndpoint` is actually a `rdzvService`. User needs to pass a k8s service with opening port. trying to understand the use case here. 

> we can automate getting the ectd server endpoint if the rdzvEndpoint field is left empty

Do you want to create an etcd server for that job or auto-discover `elastic-job` namespace to find some potential service to use? 
If that's second one, I think it's unnecessary to predicate the service if user just forget to specify it in the yaml. 

Creating job based etcd if `rdzvEndpoint` missing is a very reasonable feature request and we talked about this before but didn't add it in init version.

Reply:
@Jeffwan for rdzv; the high level idea here would be that if you are using ElasticJob controller we would hide the fact that there is etcd behind the scenes from you. <-- this would be the "tight/native" integration with kubernetes. I think there are two use-cases here:

1. shared rdzv backend; multiple jobs
   ```
    kubectl apply -k configs/sample/rdzv_backend.yaml (starts an etcd server)
    kubectl apply -k configs/sample/imagenet.yaml
   ```
2. sidecar rdzv backend with each job
    ```
    kubectl apply -k configs/sample/imagenet.yaml
    ```

The usecase for having rdzv autodiscovery when there is a shared rdzv backend, is that on a shared cluster an admin would setup the backend and the users can simply run their jobs.